migrations/inputs_cassandra/migration.go

```diff
@@ -0,0 +1,247 @@
+		}
+		if u.Hostname() == "" {
+			u.Host = "localhost:" + u.Port()
+		}
+		user := u.User.Username()
+		passwd, _ := u.User.Password()
+		key := user + ":" + passwd
+
+		endpoint, found := endpoints[key]
+		if !found {
+			endpoint = jolokiaAgent{
+				Username: user,
+				Password: passwd,
+			}
+			endpoint.fillCommon(old.InputOptions)
+		}
+		u.User = nil
+		endpoint.URLs = append(endpoint.URLs, u.String())
+		endpoints[key] = endpoint
+	}
+
+	// Create new-style metrics according to the old config
+	var javaMetrics []metricConfig
+	var cassandraMetrics []metricConfig
+	for _, metric := range old.Metrics {
+		bean := strings.TrimPrefix(metric, "/")
+
+		params := make(map[string]string)
+		parts := strings.SplitN(bean, ":", 2)
+		for _, p := range strings.Split(parts[1], ",") {
+			x := strings.SplitN(p, "=", 2)
+			params[x[0]] = x[1]
+		}
+
+		name, found := params["type"]
+		if !found {
+			return nil, fmt.Errorf("cannot determine name for metric %q", metric)
+		}
+		name = strings.SplitN(name, "/", 2)[0]
+
+		var tagKeys []string
+		var prefix *string
+		for k := range params {
+			switch k {
+			case "name", "scope", "path", "keyspace":
+				tagKeys = append(tagKeys, k)
+			}
+		}
+		sort.Strings(tagKeys)
+		for i, k := range tagKeys {
+			if k == "name" {
+				p := fmt.Sprintf("$%d_", i+1)
+				prefix = &p
+				break
+			}
+		}
+
+		switch {
+		case strings.HasPrefix(bean, "java.lang:"):
+			javaMetrics = append(javaMetrics, metricConfig{
+				Name:        name,
+				Mbean:       bean,
+				TagKeys:     tagKeys,
+				FieldPrefix: prefix,
+			})
+		case strings.HasPrefix(bean, "org.apache.cassandra.metrics:"):
+			cassandraMetrics = append(cassandraMetrics, metricConfig{
+				Name:        name,
+				Mbean:       bean,
+				TagKeys:     tagKeys,
+				FieldPrefix: prefix,
+			})
+		default:
+			return nil, fmt.Errorf("unknown java metric %q", metric)
+		}
+	}
+
+	// Create the corresponding metric configurations
+	cfg := migrations.CreateTOMLStruct("inputs", "jolokia2_agent")
+	for _, endpoint := range endpoints {
+		if len(javaMetrics) > 0 {
+			plugin := jolokiaAgent{
+				URLs:     endpoint.URLs,
+				Username: endpoint.Username,
+				Password: endpoint.Password,
+				Metrics:  javaMetrics,
+			}
+			plugin.fillCommon(old.InputOptions)
+			plugin.NamePrefix = "java"
+			cfg.Add("inputs", "jolokia2_agent", plugin)
+		}
+		if len(cassandraMetrics) > 0 {
+			plugin := jolokiaAgent{
+				URLs:     endpoint.URLs,
+				Username: endpoint.Username,
+				Password: endpoint.Password,
+				Metrics:  cassandraMetrics,
+			}
+			plugin.fillCommon(old.InputOptions)
+			plugin.NamePrefix = "cassandra"
+
+			cfg.Add("inputs", "jolokia2_agent", plugin)
+		}
+	}
+
+	// Marshal the new configuration
+	buf, err := toml.Marshal(cfg)
+	if err != nil {
+		return nil, err
+	}
+	buf = append(buf, []byte("\n")...)
+
+	// Create the new content to output
+	return buf, nil
+}
+
+func (j *jolokiaAgent) fillCommon(o common.InputOptions) {
```

Comment:
will we need to a) keep this up to date with all "common" options and b) have this in each migration? or will you generalize it out after the next one?

Reply:
I would have loved to generalize this but I ran into the issue of TOML marshalling cannot marshal an embedded struct to the same level as the embedding struct. I.e. if you have
```golang
type A struct {
    fielda string `toml:"a"
}

type B struct {
    fieldb1 string `toml:"b1"`
    fieldb2 string `toml:"b2"`
    A
}
```
you will get
``´tom
[[inputs.B]]
    b1 = "foo"
    b2 = "bar"
    [inputs.B.A]
      a = "fail"
```
instead of
 ``´tom
[[inputs.B]]
    b1 = "foo"
    b2 = "bar"
    a = "good"
```

Therefore I cannot generalize it and we need to do this in every migration or fix it in the TOML marshaller...